lib/postgrex/protocol.ex

```diff
@@ -702,7 +712,43 @@ defmodule Postgrex.Protocol do
 
     digest = :crypto.hash(:md5, [pass, user]) |> Base.encode16(case: :lower)
     digest = :crypto.hash(:md5, [digest, salt]) |> Base.encode16(case: :lower)
-    auth_send(s, msg_password(pass: ["md5", digest]), status, buffer)
+    auth_send(s, msg_password(pass: ["md5", digest, 0]), status, buffer)
+  end
+
+  defp auth_sasl(s, status = _, buffer) do
+    nonce = Enum.join(
+      for _ <- 1..@nonce_length do
+        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/"
+        |> String.codepoints
+        |> Enum.random
+      end
+    )
+    message = "n,,n=,r=" <> nonce
+    auth_send(s, msg_password(pass: ["SCRAM-SHA-256", 0, << byte_size(message) :: signed-size(32) >>, message]), status, buffer)
+  end
+
+  defp auth_cont(s, %{opts: opts} = status, data, buffer) do
```

Comment:
Can you please rewrite this function with a couple things in mind?

1. Do not use the string functions but instead use pattern matching. For example, it is fine to do `String.split/2` but given all data has the same shape, it is preferred to do this:

```
for kv <- String.split(data, ",") do
  value_size = byte_size(kv) - 2
  <<key, ?=, value::size(value_size)>> = value_size
  {key, value}
end
```

2. Avoid conversion to atoms. Let's keep the codepoint as keys, for example: `server[?s]`

3. Do not use `Enum.join` or string concatenation. Instead, we can just work with IO.data. This means that `"c=biws,r=" <> server[:r]` becomes `["c=biws,r=", server[:r]]` and that `Enum.join(["r=", server[:r], ",s=", server[:s], ",i=", server[:i]], "")` becomes simply `["r=", server[:r], ",s=", server[:s], ",i=", server[:i]]`. If for some reason you need a string, you can call `IO.iodata_to_binary`.

4. Use the `Base` module instead of `:base64`